sdk/clientcore/http-netty4/src/main/java/io/clientcore/http/netty4/implementation/Netty4ChannelInputStream.java

```diff
@@ -0,0 +1,225 @@
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) {
+        if (streamDone) {
+            return -1;
+        }
+
+        int remainingInCurrentBuffer = currentBuffer.length - readIndex;
+
+        // currentBuffer has enough data remaining to satisfy the read.
+        if (remainingInCurrentBuffer >= len) {
+            System.arraycopy(currentBuffer, readIndex, b, off, len);
+            readIndex += len;
+            return len;
+        }
+
+        // Read whatever is remaining in the current buffer.
+        System.arraycopy(currentBuffer, readIndex, b, off, remainingInCurrentBuffer);
+
+        // Set up a marker to determine how much more to read.
+        int toRead = len - remainingInCurrentBuffer;
+        off += remainingInCurrentBuffer;
+
+        while (setupNextBuffer()) {
+            if (currentBuffer.length >= toRead) {
+                // The next buffer is able to satisfy the remaining read.
+                System.arraycopy(currentBuffer, 0, b, off, toRead);
+                readIndex = toRead;
+                return len;
+            } else {
+                // The entire next buffer is able to fit in the read buffer.
+                System.arraycopy(currentBuffer, 0, b, off, currentBuffer.length);
+                toRead -= currentBuffer.length;
+                off += currentBuffer.length;
+            }
+        }
+
+        // Reached the end of the Channel contents before the requested read length could be completed.
+        // Return the amount of content read, which would be the requested length minus what couldn't be read.
+        return len - toRead;
+    }
+
+    @Override
+    public long skip(long n) {
+        if (streamDone) {
+            return 0;
+        }
+
+        int remainingInCurrentBuffer = currentBuffer.length - readIndex;
+
+        // currentBuffer has enough data remaining to satisfy the read.
+        if (remainingInCurrentBuffer >= n) {
+            readIndex += (int) n;
+            return n;
+        }
+
+        // Set up a marker to determine how much more to skip.
+        long toSkip = n - remainingInCurrentBuffer;
+
+        while (setupNextBuffer()) {
+            if (currentBuffer.length >= toSkip) {
+                // The next buffer is able to satisfy the remaining skip.
+                readIndex = (int) toSkip;
+                return n;
+            } else {
+                // The entire next buffer is able to fit in the remaining skip amount.
+                toSkip -= currentBuffer.length;
+            }
+        }
+
+        // Reached the end of the Channel contents before the skip amount could be satisfied.
+        // Return the amount of content skipped, which would be the skip amount requested minus what couldn't be
+        // skipped.
+        return n - toSkip;
+    }
+
+    @Override
+    public void close() {
+        currentBuffer = null;
+        additionalBuffers.clear();
+        if (!channelDone) {
+            channel.close();
```

Comment:
I guess this takes care of latch count on `InputStream.close` right?

Reply:
Yeah, this will end up counting down the latch